sklearn/tests/test_common.py

```diff
@@ -447,3 +447,45 @@ def test_estimators_do_not_raise_errors_in_init_or_set_params(Estimator):
 
         # Also do does not raise
         est.set_params(**new_params)
+
+
+ESTIMATOR_PARAMETER_TO_XFAIL = {
+    ("SpectralEmbedding", "affinity"),
+    ("FeatureUnion", "transformer_list"),  # Expected behavior
+    ("GaussianMixture", "weights_init"),
+    ("GaussianProcessRegressor", "alpha"),
+    ("LabelPropagation", "nn_fit"),
+    ("LabelSpreading", "nn_fit"),
+    ("NeighborhoodComponentsAnalysis", "n_components"),
+    ("PatchExtractor", "random_state"),
+    ("Pipeline", "steps"),  # Expected behavior
+    ("RANSACRegressor", "estimator"),
+    ("RidgeClassifier", "max_iter"),
+    ("RidgeClassifierCV", "alphas"),
+}
+
+
+@pytest.mark.parametrize("name, Estimator", all_estimators())
+def test_estimators_do_not_set_parameter_outside_of_init(name, Estimator):
+    """Check that parameter is not set outside of __init__."""
+    functions = chain.from_iterable(
+        (
+            f
+            for name, f in getmembers(kls, isfunction)
+            if f.__qualname__.startswith(kls.__name__) and name != "__init__"
+        )
+        for kls in Estimator.mro()
+    )
+    param_set_re = re.compile(r"self.(?!_)(\w+)(?<!_) = ")
```

Comment:
Actually, I did not think of doing such a test. I thought of calling `fit`, `predict` and friend and checking for equal `id` with original parameters.

Can we get in trouble with Cython files?

Reply:
If `getsource` was called on a Cython method, then `getsource` would break. For our case, calling `getmembers(kls, isfunction)` on the class returns unbounded functions for Python classes, and bounded Cython methods.

For example, for a Cython class:

```python
from sklearn.metrics._dist_metrics import DistanceMetric
import inspect

dict(inspect.getmembers(DistanceMetric))["dist_to_rdist"]
<method 'dist_to_rdist' of 'sklearn.metrics._dist_metrics.DistanceMetric' objects>
```

but for a Python class:

```python
from sklearn.tree import DecisionTreeClassifier
import inspect

dict(inspect.getmembers(DecisionTreeClassifier))["fit"]
<function sklearn.tree._classes.DecisionTreeClassifier.fit(self, X, y, sample_weight=None, check_input=True)>
```

I can be extra safe and catch the `TypeError` in `getsource` and continue.

In general, parsing source code like this can be brittle. This one is not too bad since it only checks for `self.some_word = ` in the source, which seems to work in catching our error cases.

Reply:
``r"self.(?!_)(\w+)(?<!_) = "``
don't you need to escape the "." ? (very unlikely to trigger some false positive though)